### PR TITLE
Fix: gateway approval card gets a usable approval_id (#6008)

### LINKED ISSUE #6008: Follow-up to #5269: approval card is displayed, but approval_id is empty and approval cannot be submitted
Hello,

**Summary**

This is a follow-up to #5269.

The original problem in #5269 was that no approval dialog or approval card was displayed at all. That part now appears to be fixed: the WebUI displays the approval request.

The remaining problem is that the displayed approval request has an empty approval_id. When the user tries to approve or deny it, the WebUI cannot submit the response. The approval request is visible, but it is not actionable.

Image

**Environment**

    Infrastructure: Oracle VPS Free Tier
    OS: Ubuntu 24.04 ARM
    Host resources: 4 vCPU, 24 GB RAM
    Container runtime: Podman, rootless
    Deployment: separate hermes-agent and hermes-webui containers
    Hermes WebUI: v0.52.0
    Hermes Agent: v0.18.2
    WebUI chat backend: gateway
    Gateway status in WebUI: Running
    Runs API opt-in: HERMES_WEBUI_GATEWAY_USE_RUNS_API=true

Relevant non-secret configuration:

```
HERMES_WEBUI_CHAT_BACKEND=gateway
HERMES_WEBUI_GATEWAY_BASE_URL=http://<hermes-agent-host>:8642
HERMES_WEBUI_GATEWAY_USE_RUNS_API=true
```

The API server key and the WebUI gateway API key are supplied through environment files. Their values are not included here.

The same Hermes Agent can also be used through the separate Hermes Dashboard. The approval problem described here has not been reproduced through that Dashboard path.

**Steps to reproduce**

    Run Hermes Agent with its API server enabled.
    Connect Hermes WebUI to the remote gateway/API server.
    Use the gateway-backed WebUI chat path with the Runs API enabled.
    Ask the agent to perform an operation that requires approval, such as a guarded terminal or file operation.
    Wait for the approval request.
    Observe that the approval card or request is displayed.
    Click an approval choice such as Allow once or Deny.

**Expected behavior**

The WebUI should assign or receive a non-empty identifier for the approval request and submit the selected choice to the Hermes Agent Runs API.

The approval should be resolved through:

`POST /v1/runs/{run_id}/approval`

The run should then continue for an allowed choice or stop cleanly for a denied choice.

**Actual behavior**

The approval request is visible, but the WebUI-side approval object has an empty approval_id.

The following represents the normalized approval object observed in the WebUI path. It is not intended to represent the raw SSE payload:

```
{
  "event": "approval.request",
  "run_id": "<redacted-run-id>",
  "approval_id": ""
}

```
The exact run ID and timestamp are omitted because they are session-specific.

When the user submits the approval choice, the WebUI cannot complete the response because the gateway approval response path requires a non-empty approval_id.

The choice therefore cannot be validated from the browser, and the run remains blocked.

**Source evidence**

In WebUI v0.52.0, api/gateway_chat.py translates the identifier as follows:

`approval_id = str(payload.get("approval_id") or payload.get("id") or "").strip()`

The translated approval object contains this value as approval_id. If the upstream event contains neither approval_id nor id, the resulting value is the empty string.

In api/routes.py, the gateway response path contains this guard:

```
if _run_id:
    if not approval_id:
        return bad(handler, "approval_id is required for gateway approvals")
```

Therefore, an approval event without an identifier can be displayed but cannot pass through the WebUI approval response endpoint.

Relevant WebUI source:

    https://github.com/nesquena/hermes-webui/blob/v0.52.0/api/gateway_chat.py
    https://github.com/nesquena/hermes-webui/blob/v0.52.0/api/routes.py
    https://github.com/nesquena/hermes-webui/blob/v0.52.0/api/runner_client.py

**Agent/API contract observed in source**

The Hermes Agent API-server Runs API emits an _approval.request_ event containing the run identifier and approval metadata.

In gateway/platforms/api_server.py, the event builder adds fields including:

```
{
    "event": "approval.request",
    "run_id": run_id,
    "timestamp": time.time(),
    "choices": ["once", "session", "always", "deny"],
}
```

The event builder does not add an approval_id field in this path.

The corresponding Hermes Agent endpoint is:

`POST /v1/runs/{run_id}/approval`

Its handler uses the URL run_id to retrieve the active approval session and calls:

```
resolve_gateway_approval(
    approval_session_key,
    choice,
    resolve_all=resolve_all,
)
```

The request body is validated for choice. The submitted approval_id is not used by that handler to identify the pending approval.

Relevant Hermes Agent source:

    https://github.com/NousResearch/hermes-agent/blob/main/gateway/platforms/api_server.py

The API-server approval functionality was introduced as the resolution of Hermes Agent issue #15802 through PR #21899:

    https://github.com/NousResearch/hermes-agent/issues/15802
    https://github.com/NousResearch/hermes-agent/pull/21899

This report is narrower than #15802. The Runs API now exposes an approval event and a run-scoped resolution endpoint, but the WebUI adapter still produces an empty local identifier when the event does not contain one.

**Historical context**

The original issue #5269 reported that no approval request dialog was displayed. It used this deployment shape and reported errors such as:

`Tool write_file returned error: Write denied: '<path>' is a protected system/credential file.`

Issue #5269 was closed through PR #5640. The PR description states that the release batch contained documentation, environment examples, and tests, with zero application-code changes. The relevant gateway approval documentation was included through PR #5549.

Links:

    Previous issue: https://github.com/nesquena/hermes-webui/issues/5269
    Closing release PR: https://github.com/nesquena/hermes-webui/pull/5640
    Gateway approval documentation PR: https://github.com/nesquena/hermes-webui/pull/5549

This new issue reports a later state: the approval card is now visible, but the response cannot be submitted because the local approval_id is empty.

Related approval follow-ups include #4203, #4479, and #4771:

    https://github.com/nesquena/hermes-webui/issues/4203
    https://github.com/nesquena/hermes-webui/issues/4479
    https://github.com/nesquena/hermes-webui/issues/4771

**Likely interoperability gap**

The two sides currently use different identifiers for different purposes:

    Hermes Agent resolves the pending approval from the run-scoped run_id and its server-side approval session.
    Hermes WebUI requires a non-empty approval_id for its local approval-card and response state.
    The Runs API approval event does not provide the WebUI-specific identifier.

This produces the following sequence:

    Hermes Agent emits approval.request with run_id and no approval_id.
    Hermes WebUI translates the missing value to approval_id == "".
    The WebUI displays the approval request.
    /api/approval/respond rejects the response before relaying it because approval_id is empty.
    The Hermes Agent run remains waiting for approval.

**Proposed fix**

The WebUI Runs API adapter should not expose an empty approval identifier to the approval UI.

One possible WebUI-side fix is:

    Preserve a gateway-provided approval_id when present.
    Otherwise preserve a gateway-provided id when present.
    Otherwise generate a non-empty WebUI-local correlation ID from the event identity, for example run_id, event timestamp, command, and pattern keys.
    Keep the generated value stable when the same SSE event is replayed or translated again.
    Continue using the Agent-provided run_id for the actual Runs API resolution request.

The generated identifier would be a WebUI correlation key, not a replacement for the Hermes Agent run identifier. Hermes Agent should remain responsible for resolving the pending approval by run_id.

An alternative would be to make the WebUI response path accept a missing local approval_id for Runs API approvals and route solely by run_id, but that would require careful handling of stale and concurrent approval cards. A local deterministic correlation ID appears to preserve the existing WebUI response contract with a smaller behavior change.

**Suggested regression tests**

Please consider adding tests that verify:

    A Runs API approval.request payload without approval_id or id is translated to an approval object with a non-empty local identifier.
    Translating the same payload twice produces the same local identifier.
    A gateway-provided approval_id is preserved unchanged.
    A gateway-provided id is used when approval_id is absent.
    The generated identifier is accepted by /api/approval/respond.
    The choice is relayed to POST /v1/runs/{run_id}/approval.
    The request body sent to Hermes Agent contains the choice and does not rely on the local identifier for server-side approval lookup.
    Concurrent approval requests do not accidentally share the same WebUI correlation identifier.

**Verification status**

The empty-ID behavior is confirmed by source inspection and by the observed WebUI behavior described above.

No claim is made here that a proposed patch has passed the WebUI test suite. A checkout of the exact target branch and a controlled end-to-end run would still be needed to validate the implementation and check behavior with concurrent or replayed approval events.

**Maintainer question**

Could you confirm whether the intended WebUI contract for Runs API approvals is:

    a WebUI-generated local correlation ID combined with the Agent run_id, or
    routing entirely by the Agent run_id without requiring a local approval_id?

The current v0.52.0 code appears to require both for the browser response path, while the Hermes Agent Runs API resolves the approval using run_id and the active approval session.

This version is deliberately cautious: it does not claim that the raw Agent event contains "approval_id": ""; it states that the Agent event omits the field and that the WebUI normalizes the missing value to an empty string.

### PR DESCRIPTION
**Pull request by rudidev08**
Closes #6008

## Thinking Path

- Hermes WebUI aims for near 1:1 parity with the Hermes CLI in a browser, including the approval gate for risky agent actions.
- When the WebUI runs chat through the Gateway runs API, the agent's `approval.request` SSE event carries `run_id` but no `approval_id`.
- `_gateway_runs_approval_event` normalized the missing field to `""`, and the runs-API path streams the card straight to the frontend — it never passes through `submit_pending`, whose `setdefault("approval_id", uuid)` would have filled one in.
- With an empty id the card renders but `_handle_approval_respond` rejects the response with 400, so the run stays blocked forever.
- This PR synthesizes a WebUI-local `approval_id` at the one translation point both gateway paths share, and records `approval_id -> run_id` server-side so the response still relays after the stream worker exits.

## What Changed

- `api/gateway_chat.py`
  - `_gateway_runs_approval_event`: when the payload has no `approval_id` and no `id`, synthesize `"gw-" + uuid4().hex` — unique per event.
  - New `_APPROVAL_RUN_IDS` map: `(session_id, approval_id) -> run_id`, capped at 256 entries, populated in both stream loops (runs-API and legacy) when an approval card is emitted, popped on successful relay.
  - Replaced `import hashlib` with `import uuid`.
  - Upstream identifiers are still preferred: `approval_id`, then `id`, then the synthesized fallback.
- `api/routes.py` — `_handle_approval_respond`
  - When `active_stream_id` is gone, recover the run_id from `_APPROVAL_RUN_IDS` before the existing gateway-mirror fallback.
  - Pop the entry after a successful relay.
- `tests/test_gateway_approval_runs_api.py`
  - `id`-only payload uses `id` verbatim (no synthesis).
  - New: payload with neither field gets a non-empty `gw-` id; two translations of the same payload get distinct ids (per-event uniqueness).
  - New: end-to-end relay of a synthesized id returns 200.
  - New: relay recovers the run_id from `_APPROVAL_RUN_IDS` when the stream pointer is gone, and retires the mapping on success.
- `tests/test_gateway_approval_legacy_path.py`
  - Widened the sse_event-reset scan window (1500 -> 1700) to cover the added registration line.

## Why It Matters

- Restores the ability to approve or deny gateway-run agent actions from the WebUI; without it the card is a dead end and the run never resumes.
- A non-empty unique id is what the frontend's dismissed-card guard and render-dedupe signature require, and what `_handle_approval_respond` needs to route the response.
- Per-event uniqueness matters: a repeat approval for the same command in one run must get its own id, or dismissing the first card would make the frontend suppress the second and re-block the run.
- The `_APPROVAL_RUN_IDS` map is needed because `_STREAM_RUN_IDS` is popped when the stream worker exits (e.g. a run parked on approval past the SSE read timeout) and the gateway pending mirror never holds runs-API approvals — so without it a click on a still-rendered card finds no run_id and fails with `ok:false`.

## Maintainer-guidance note

The issue thread carries two suggested fixes:

1. Synthesize a WebUI-local identifier in `_gateway_runs_approval_event`.
2. Drop the empty-identifier requirement and route by `run_id` alone.

This PR implements option 1. Reasons from the code:

- The frontend needs a non-empty unique `approval_id` for its dismissed-card guard and render-dedupe signature.
- `_resolve_approval_legacy` with an empty id pops the oldest pending entry, which could consume an unrelated local pending approval — the #527 parallel-approval guard exists against exactly this.
- The mirror-based `run_id` recovery in `_handle_approval_respond` needs a non-empty id.

Happy to switch to the route-by-`run_id` variant if you prefer it.

## Known limitation: parallel approvals in one run

- The gateway resolves approvals FIFO and ignores the posted `approval_id` (`resolve_gateway_approval` in hermes-agent `tools/approval.py`).
- With two approvals pending in one run (parallel subagents, or a batch of approval-gated commands), the card shows the newest command but Approve resolves the oldest — a command the user never explicitly saw.
- Scope is per chat session: each session has its own queue, so sequential tool use never hits it. I could not reproduce it in practice.
- Pre-existing for payloads with real ids; this fix widens exposure only by making id-less cards clickable at all.
- Proper fix belongs in hermes-agent: match on `approval_id` when supplied, falling back to FIFO. The WebUI already posts the id on every respond, so no further WebUI change is needed once the agent honors it.

## Verification

- State/event surface: gateway approval event translation, the `approval_id -> run_id` recovery map, and the approval response relay (runs-API + legacy paths).
- `./scripts/test.sh tests/test_gateway_approval_runs_api.py tests/test_gateway_approval_legacy_path.py` — 35 passed.
- `./scripts/test.sh -k approval` — 267 passed, 4 skipped.
- `ruff check` on the changed lines — clean (pre-existing findings elsewhere in `api/routes.py` are untouched).
- Manually tested by a human contributor and confirmed working. Note: the contributor is not the original reporter and was not affected by this bug directly, so the reporter's exact pre-fix condition could not be reproduced first-hand.

## Risks / Follow-ups

- The synthesized id is opaque and WebUI-local; the agent ignores it and resolves the parked run FIFO by the URL `run_id`, so it never needs to match anything upstream.
- Parallel-approval limitation above needs a hermes-agent change to fully close.
- If the maintainer prefers routing by `run_id` alone (option 2), the frontend guard and `_resolve_approval_legacy` empty-id behavior would need adjustment too — out of scope here.
- No UI change: the card already renders; a non-empty id makes the existing submit path work, so no before/after screenshots apply.

## Model Used

- Provider: Anthropic
- Model: Claude Opus 4.8 (`claude-opus-4-8`), via Claude Code CLI.
- AI produced the initial synthesized-id fix and tests. The contributor revised the identifier to a per-event uuid, added the server-side `(session_id, approval_id) -> run_id` recovery map and the respond-endpoint lookup, and manually verified the fix.

### CODE REVIEW

**Comment by rudidev08** [comment]
Note: claude was warning me that this fix and overall implementation works for single tool approval. If there are multiple tools needing approval in same time, approving one might approve the other.

I was not able to reproduce this locally, since I keep getting one tool approval at a time, so i don't know how much above is an actual issue. I'll see about opening issue about it just in case.

File: api/gateway_chat.py [hunk 1]
```diff
@@ -38,6 +39,25 @@
...
+_APPROVAL_RUN_IDS: dict[tuple[str, str], str] = {}
+_APPROVAL_RUN_IDS_MAX = 256
+
+
+def _register_approval_run_id(session_id: str, approval_id: str, run_id: str) -> None:
+    if not (session_id and approval_id and run_id):
+        return
+    _APPROVAL_RUN_IDS[(session_id, approval_id)] = run_id
+    while len(_APPROVAL_RUN_IDS) > _APPROVAL_RUN_IDS_MAX:
+        del _APPROVAL_RUN_IDS[next(iter(_APPROVAL_RUN_IDS))]
```
[the remark this answers is not in the document]
Comment (rudidev08) [reply api/gateway_chat.py]: Leaving this as is. Eviction only starts past 256 approvals that were never answered, accumulated process-wide, and the evicted entry only matters if that card's stream worker has also exited. A single-user WebUI doesn't realistically reach that state. The cap exists only to bound growth from abandoned approvals; skipping entries with a live stream would add state tracking for a condition that doesn't occur in practice.

[the remark this answers is not in the document]
Comment (rudidev08) [reply api/gateway_chat.py]: Fixed in be2a659b: eviction and the respond endpoint's pop now share a lock.

**Comment by rudidev08** [comment]
Re the summary finding "Active Stream Overrides Approval Mapping": fixed in be2a659b. `_handle_approval_respond` now prefers the card's own `(session_id, approval_id) -> run_id` mapping over the session's active stream, so a stale card's choice relays to its own run instead of whatever run is currently active. The active-stream and mirror fallbacks are kept for the post-restart case, and a regression test covers the stale-card path.

**Comment by rudidev08** [comment]
above mentioned parallel issue is actually with hermes agent, i've filed issue with their repo: https://github.com/NousResearch/hermes-agent/issues/64001 (doesn't block this)

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
## 🔬 Gate review (Codex, gateway approval safety-gate) — SHIP ONLY WITH FIXES

Thanks @rudidev08 — you correctly diagnosed #6008 (the Runs-API `approval.request` SSE carries `run_id` but an empty `approval_id`, so the card can't be submitted), and the happy path works: the ordinary empty-id approve-once and deny round-trips send the exact displayed id, `deny` reaches the agent as a real block, and tuple keying isolates identical ids across sessions. But the synthesized-id-map approach introduced 4 blockers on this safety-gate path:

1. **[CORE — wrong-run routing] `api/routes.py:23445`** — an explicit id can fall back to an unrelated active run (a stale card-A could resolve active-run-B). Make approval-specific ownership authoritative; prohibit explicit ids from falling back to an unrelated active run.
2. **[CORE — unlocked map: live-eviction + concurrent KeyError] `api/gateway_chat.py:54`** — the bounded map (`while len>MAX: del next(iter)`) isn't lock-guarded: a burst can evict a still-**unresolved** approval (re-breaking submit, with a false `200 stale_cleared`), and concurrent registration throws `KeyError` (reproduced). Add locked map helpers + a live-safe capacity policy that never evicts unresolved approvals.
3. **[CORE — unstable id synthesis] `api/gateway_chat.py:370`** — a UUID minted per translation means the same event can yield different ids (card shows one, submit uses another). Use a stable event-derived id.
4. **[CORE — no terminal cleanup] `api/gateway_chat.py:553`** — completed/failed/cancelled runs don't retire their approval mappings (only `_STREAM_RUN_IDS` is popped), so stale entries fill the cap and trigger the live eviction above. Clean on confirmed terminal lifecycle events (retain only for transport-only loss where the remote run may still await approval).

Also the new tests manually seed the map / `_STREAM_RUN_IDS` rather than driving a real empty-id SSE → registration → approve/deny, so they miss all of the above.

**Note (maintainer + both #6008 authors):** the competing #6040 (mirror-reuse, @rodboev) is architecturally cleaner and avoids blockers 2-4, but it *also* kicks back because — like this PR at a different layer — **the run_id↔approval correlation record isn't created at gateway-request emission time**, so the Runs-API path has nothing durable+stable to correlate against. I'm flagging the #6008 approach to @nesquena: the robust fix likely wants a maintainer-blessed correlation design (create a stable record at gateway emission → carry run_id on submit → match on response → retire on terminal), and then whichever PR implements *that* ships. Great diagnosis; the shape just needs the correlation established upstream rather than reconstructed at submit.

**Comment by nesquena-hermes** [comment]
Thanks @rudidev08. This is now superseded by the canonical #6040 implementation, which shipped in `exp-v0.52.147`; the shared tracker #6008 is closed as completed.

Your PR independently identified the missing WebUI-local approval correlation and terminal-cleanup requirements. Those findings were useful in the bake-off and were carried into the maintainer correlation contract before #6040 converged.

Closing this competing implementation so the shipped fix remains the single maintained lane. Credit for the diagnosis and iteration is preserved here.
